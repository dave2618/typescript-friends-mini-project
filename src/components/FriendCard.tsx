import React, { useContext } from 'react';
import { useNavigate } from 'react-router-dom';
import Card from '@mui/material/Card';
import CardContent from '@mui/material/CardContent';
import CardMedia from '@mui/material/CardMedia';
import Typography from '@mui/material/Typography';
import Button from '@mui/material/Button';
import ThumbUpIcon from '@mui/icons-material/ThumbUp';
import ThumbDownIcon from '@mui/icons-material/ThumbDown';
import { CardActionArea } from '@mui/material';

import { FriendsType } from '../types/FriendsType';
import FriendsContext from '../contexts/FriendsContext';

interface Props { 
    person: FriendsType,
}

const FriendCard = (props: Props) => {

    const person: FriendsType = props.person;
    const { friends, setFriends } = useContext(FriendsContext);
    const navigate = useNavigate();

    const like = () => () => {
        const objIndex = friends.findIndex((obj => obj.name === person.name));
        setFriends((prev: Array<FriendsType>) => {
            const temp = [...prev];
            ++temp[objIndex].likes;
            return temp;
        });
    }

    const dislike = () => () => {
        if(person.likes > 0) {
            const objIndex = friends.findIndex((obj => obj.name === person.name));
            setFriends((prev: Array<FriendsType>) => {
                const temp = [...prev];
                --temp[objIndex].likes;
                return temp;
            });
        }
    }

    const handleClick = (personData: FriendsType) => {
        navigate('/details', { state: personData });
    }

    return (
        <div>
            <Card sx={{ maxWidth: 500 }}>
                <CardActionArea onClick={() => handleClick(person)}>
                <CardMedia
                    component="img"
                    height="200"
                    image={person.image}
                    alt={person.name}
                    data-testid="PersonImage"
                />
                </CardActionArea>
                <CardContent>
                    <Typography gutterBottom variant="h6" component="div">
                        {person.name}
                    </Typography>
                    
                <p data-testid='LikeLabel' style={{fontWeight: 'bold'}}>Likes: {person.likes}</p>

                <Button data-testid='LikeButton' variant="outlined" onClick={like()} style={{marginRight: '10px'}}>
                    <span style={{marginTop: '2px'}}>Like</span>
                    <ThumbUpIcon style={{marginLeft: '7px'}} fontSize='small' />
                </Button>

                <Button data-testid='DisLikeButton' variant="outlined" onClick={dislike()}>
                    <span style={{marginTop: '2px'}}>Dislike</span>
                    <ThumbDownIcon style={{marginLeft: '7px'}} fontSize='small' />
                </Button>
                </CardContent>
            </Card>
        </div>
    );
}

export default FriendCard;