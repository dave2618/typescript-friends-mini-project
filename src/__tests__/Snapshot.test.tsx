import { BrowserRouter } from 'react-router-dom';
import FriendCard from '../components/FriendCard';
import TestRenderer from 'react-test-renderer';

test('match the snapshot', () => {
    const person = {
            name: 'Andrew',
            city: 'London',
            job: 'Account Manager',
            phone: '0753212345',
            image: 'https://loremflickr.com/640/480/people?72705',
            likes: 0,
    }
    const tree = TestRenderer.create(<BrowserRouter><FriendCard person={person} /> </BrowserRouter>).toJSON();
    expect(tree).toMatchSnapshot();
});
