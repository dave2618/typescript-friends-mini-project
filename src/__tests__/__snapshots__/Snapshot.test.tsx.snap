// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`match the snapshot 1`] = `
Array [
  <div>
    <div
      className="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiCard-root css-1gcg9uy-MuiPaper-root-MuiCard-root"
    >
      <button
        className="MuiButtonBase-root MuiCardActionArea-root css-b6lsxj-MuiButtonBase-root-MuiCardActionArea-root"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onContextMenu={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
      >
        <img
          alt="Andrew"
          className="MuiCardMedia-root MuiCardMedia-media MuiCardMedia-img css-o69gx8-MuiCardMedia-root"
          data-testid="PersonImage"
          height="200"
          src="https://loremflickr.com/640/480/people?72705"
        />
        <span
          className="MuiCardActionArea-focusHighlight css-1v2exvi-MuiCardActionArea-focusHighlight"
        />
      </button>
      <div
        className="MuiCardContent-root css-46bh2p-MuiCardContent-root"
      >
        <div
          className="MuiTypography-root MuiTypography-h6 MuiTypography-gutterBottom css-18k87ye-MuiTypography-root"
        >
          Andrew
        </div>
        <p
          data-testid="LikeLabel"
          style={
            Object {
              "fontWeight": "bold",
            }
          }
        >
          Likes: 
          0
        </p>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          data-testid="LikeButton"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          style={
            Object {
              "marginRight": "10px",
            }
          }
          tabIndex={0}
          type="button"
        >
          <span
            style={
              Object {
                "marginTop": "2px",
              }
            }
          >
            Like
          </span>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="ThumbUpIcon"
            focusable="false"
            style={
              Object {
                "marginLeft": "7px",
              }
            }
            viewBox="0 0 24 24"
          >
            <path
              d="M1 21h4V9H1v12zm22-11c0-1.1-.9-2-2-2h-6.31l.95-4.57.03-.32c0-.41-.17-.79-.44-1.06L14.17 1 7.59 7.59C7.22 7.95 7 8.45 7 9v10c0 1.1.9 2 2 2h9c.83 0 1.54-.5 1.84-1.22l3.02-7.05c.09-.23.14-.47.14-.73v-2z"
            />
          </svg>
        </button>
        <button
          className="MuiButtonBase-root MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium MuiButton-root MuiButton-outlined MuiButton-outlinedPrimary MuiButton-sizeMedium MuiButton-outlinedSizeMedium css-1rwt2y5-MuiButtonBase-root-MuiButton-root"
          data-testid="DisLikeButton"
          disabled={false}
          onBlur={[Function]}
          onClick={[Function]}
          onContextMenu={[Function]}
          onDragLeave={[Function]}
          onFocus={[Function]}
          onKeyDown={[Function]}
          onKeyUp={[Function]}
          onMouseDown={[Function]}
          onMouseLeave={[Function]}
          onMouseUp={[Function]}
          onTouchEnd={[Function]}
          onTouchMove={[Function]}
          onTouchStart={[Function]}
          tabIndex={0}
          type="button"
        >
          <span
            style={
              Object {
                "marginTop": "2px",
              }
            }
          >
            Dislike
          </span>
          <svg
            aria-hidden={true}
            className="MuiSvgIcon-root MuiSvgIcon-fontSizeSmall css-ptiqhd-MuiSvgIcon-root"
            data-testid="ThumbDownIcon"
            focusable="false"
            style={
              Object {
                "marginLeft": "7px",
              }
            }
            viewBox="0 0 24 24"
          >
            <path
              d="M15 3H6c-.83 0-1.54.5-1.84 1.22l-3.02 7.05c-.09.23-.14.47-.14.73v2c0 1.1.9 2 2 2h6.31l-.95 4.57-.03.32c0 .41.17.79.44 1.06L9.83 23l6.59-6.59c.36-.36.58-.86.58-1.41V5c0-1.1-.9-2-2-2zm4 0v12h4V3h-4z"
            />
          </svg>
        </button>
      </div>
    </div>
  </div>,
  " ",
]
`;
